frame/grandpa/src/migrations.rs

```diff
@@ -15,5 +15,51 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use frame_support::{
+	traits::{Get, OnRuntimeUpgrade},
+	weights::Weight,
+};
+
+use crate::{Config, CurrentSetId, SetIdSession, LOG_TARGET};
+
 /// Version 4.
 pub mod v4;
+
+/// This migration will clean up all stale set id -> session entries from the
+/// `SetIdSession` storage map, only the latest `max_set_id_session_entries`
+/// will be kept.
+///
+/// This migration should be added with a runtime upgrade that introduces the
+/// `MaxSetIdSessionEntries` constant to the pallet (although it could also be
+/// done later on).
+pub struct CleanupSetIdSessionMap<T>(sp_std::marker::PhantomData<T>);
+impl<T: Config> OnRuntimeUpgrade for CleanupSetIdSessionMap<T> {
+	fn on_runtime_upgrade() -> Weight {
+		// NOTE: since this migration will loop over all stale entries in the
+		// map we need to set some cutoff value, otherwise the migration might
+		// take too long to run. for scenarios where there are that many entries
+		// to cleanup a multiblock migration will be needed instead.
+		if CurrentSetId::<T>::get() > 25_000 {
```

Comment:
Just curious.
In case we have more than `X` entries this will bail out entirely.

In this case what is the strategy to cleanup the leftovers?
The note talks about a multi-block migration. How this can be implemented? And who is in charge to implement it?  Just asking to prevent it from being forgotten forever 😃

Also at this point can't we at least use a best effort approach and cleanup the first `X` entries?

Another idea/question (not a super expert of runtime upgrades... but I assume it can work)
If we save in the state a temporary variable to keep track of where we arrived (with the cleanup) so far  and we leave this migration code here for the next runtime, then in the worst case after N upgrades we cleaned everything and at that point we also remove the temporary state variable.

Reply:
I just put this (random) cut-off here to prevent any users to shoot themselves in the foot. By "multi-block" migration I mean exactly what you suggested, just keep track of the current state of cleanup and remove N entities every block until it's finished. The reason I didn't implement this is because we don't need it for any of our networks and I figured it was unnecessary to add the extra complexity. I put this note here in case any other user of substrate needs to implement this.

Reply:
Got it! Thank you